File: framework/languages/dotnet/e2e/RegistryMessaging/Client/Scenarios/RmB2ScaleInScenario.cs
using Zlink.HttpClient;
using RegistryMessaging.Client;
using RegistryMessaging.Shared;

namespace RegistryMessaging.Client.Scenarios;

// RM-B2 verifies that traffic continues through the remaining provider after
// one provider is stopped for scale-in.
internal static class RmB2ScaleInScenario
{
    public static async Task RunAsync(ClientOptions options)
    {
        await using var cluster = await DynamicClusterLauncher.StartAsync(options, "rm-b2");
        var providerA = await cluster.StartProviderAsync("api-a", "api-a");
        var providerB = await cluster.StartProviderAsync("api-b", "api-b");
        using var requester = ZLinkHttpClient.Create(providerA.HttpUrl)
            .Json()
            .Timeout(TimeSpan.FromMinutes(5))
            .Build();
        using var providerBClient = ZLinkHttpClient.Create(providerB.HttpUrl)
            .Json()
            .Timeout(TimeSpan.FromMinutes(5))
            .Build();

        var beforeA = await ReadEvidenceAsync(requester);
        var beforeB = await ReadEvidenceAsync(providerBClient);
        var markerBefore = $"rm-b2-before-{Guid.NewGuid():N}";
        var valuesBefore = Enumerable.Range(0, 40)
            .Select(index => $"{markerBefore}-{index}")
            .ToArray();
        var repliesBefore = new List<ProfileReply>(valuesBefore.Length);
        foreach (var value in valuesBefore)
        {
            var reply = (await requester.Post("/profile/request")
                .Body(new ProfileRequest(value))
                .SubmitAsync<ProfileReply>()).Body;
            repliesBefore.Add(reply);
        }

        ScenarioAssert.That(repliesBefore.Count == valuesBefore.Length, "RM-B2 pre-scale reply count mismatch.");
        ScenarioAssert.That(
            repliesBefore.All(reply => reply.ProviderRid is "api-a" or "api-b"),
            "RM-B2 reply provider mismatch before scale-in.");

        var apiABeforeValues = repliesBefore
            .Where(reply => reply.ProviderRid == "api-a")
            .Select(reply => reply.Value)
            .ToArray();
        var apiBBeforeValues = repliesBefore
            .Where(reply => reply.ProviderRid == "api-b")
            .Select(reply => reply.Value)
            .ToArray();
        ScenarioAssert.That(apiABeforeValues.Length > 0 && apiBBeforeValues.Length > 0,
            "RM-B2 expected both providers before scale-in.");
        var scaleOutA = await WaitEvidenceAsync(requester, apiABeforeValues[^1]);
        var scaleOutB = await WaitEvidenceAsync(providerBClient, apiBBeforeValues[^1]);
        var preA = ScenarioAssert.CountNewEvidence(scaleOutA, beforeA, "profile-request|rid=api-a", markerBefore);
        var preB = ScenarioAssert.CountNewEvidence(scaleOutB, beforeB, "profile-request|rid=api-b", markerBefore);
        ScenarioAssert.That(preA == apiABeforeValues.Length && preB == apiBBeforeValues.Length
            && preA + preB == valuesBefore.Length, "RM-B2 expected both providers before scale-in.");

        await cluster.StopAsync(providerB);
        await Task.Delay(TimeSpan.FromSeconds(1));

        beforeA = await ReadEvidenceAsync(requester);
        beforeB = await ReadEvidenceIgnoringStoppedAsync(providerBClient);
        var markerAfter = $"rm-b2-after-{Guid.NewGuid():N}";
        var valuesAfter = Enumerable.Range(0, 20)
            .Select(index => $"{markerAfter}-{index}")
            .ToArray();
        var repliesAfter = new List<ProfileReply>(valuesAfter.Length);
        foreach (var value in valuesAfter)
        {
            var reply = (await requester.Post("/profile/request")
                .Body(new ProfileRequest(value))
                .SubmitAsync<ProfileReply>()).Body;
            repliesAfter.Add(reply);
        }

        ScenarioAssert.That(repliesAfter.Count == valuesAfter.Length, "RM-B2 post-scale reply count mismatch.");
        ScenarioAssert.That(
            repliesAfter.All(reply => reply.ProviderRid == "api-a"),
            "RM-B2 after scale-in should reach api-a only.");

        var afterA = await WaitEvidenceAsync(requester, valuesAfter[^1]);
        var afterB = await ReadEvidenceIgnoringStoppedAsync(providerBClient);
        var a = ScenarioAssert.CountNewEvidence(afterA, beforeA, "profile-request|rid=api-a", markerAfter);
        var b = ScenarioAssert.CountNewEvidence(afterB, beforeB, "profile-request|rid=api-b", markerAfter);
        ScenarioAssert.That(a == valuesAfter.Length && b == 0, "RM-B2 expected only api-a after scale-in.");

        Console.WriteLine("scenario RM-B2 passed");
    }

    static async Task<string[]> ReadEvidenceAsync(ZLinkHttpClient http) =>
        (await http.Get("/evidence").SubmitAsync<string[]>()).Body;

    static async Task<string[]> ReadEvidenceIgnoringStoppedAsync(ZLinkHttpClient http)
    {
        try
        {
            return await ReadEvidenceAsync(http);
        }
        catch
        {
            return [];
        }
    }

    static async Task<string[]> WaitEvidenceAsync(ZLinkHttpClient http, string contains) =>
        (await http.Post("/evidence/wait")
            .Body(new EvidenceWaitRequest(contains))
            .SubmitAsync<string[]>()).Body;
}

File: framework/languages/node/e2e/RegistryMessaging/Client/Scenarios/rm-b2-scale-in-scenario.ts
import type { ClientOptions } from '../Support/client-options';
import { DynamicClusterLauncher } from '../Support/dynamic-cluster-launcher';
import { getJson, postJson } from '../Support/http-client';
import { countNewEvidence, ensure, uniqueMarker } from '../Support/scenario-assert';
import type { ProfileRes } from '../../Shared/messages';

export async function runRmB2(options: ClientOptions): Promise<void> {
  const cluster = await DynamicClusterLauncher.start(options, 'rm-b2');
  try {
    const providerA = await cluster.startProvider('api-a', 'api-a');
    const providerB = await cluster.startProvider('api-b', 'api-b');
    const consumer = await cluster.startConsumer('rm-b2-consumer');
    await cluster.waitForProviders([providerA.channelEndpoint, providerB.channelEndpoint]);
    const markerBefore = uniqueMarker('rm-b2-before');
    const beforeA = await getJson<string[]>(providerA.httpUrl, '/evidence');
    const beforeB = await getJson<string[]>(providerB.httpUrl, '/evidence');
    const repliesBefore: Array<{ requestValue: string; reply: ProfileRes }> = [];
    for (let i = 0; i < 40; i += 1) {
      const requestValue = `${markerBefore}-${i}`;
      repliesBefore.push({
        requestValue,
        reply: await postJson<ProfileRes>(consumer.httpUrl, '/profile/request', { value: requestValue })
      });
    }
    const apiABeforeValues = repliesBefore.filter((entry) => entry.reply.providerRid === 'api-a').map((entry) => entry.requestValue);
    const apiBBeforeValues = repliesBefore.filter((entry) => entry.reply.providerRid === 'api-b').map((entry) => entry.requestValue);
    ensure(apiABeforeValues.length > 0 && apiBBeforeValues.length > 0, 'RM-B2 expected both providers before scale-in.');
    const scaleOutA = await postJson<string[]>(providerA.httpUrl, '/evidence/wait', { contains: apiABeforeValues[apiABeforeValues.length - 1] });
    const scaleOutB = await postJson<string[]>(providerB.httpUrl, '/evidence/wait', { contains: apiBBeforeValues[apiBBeforeValues.length - 1] });
    ensure(
      countNewEvidence(scaleOutA, beforeA, 'profile-request|rid=api-a', markerBefore)
      + countNewEvidence(scaleOutB, beforeB, 'profile-request|rid=api-b', markerBefore) === 40,
      'RM-B2 expected both providers before scale-in.'
    );

    const traffic = sendContinuously(consumer.httpUrl, uniqueMarker('rm-b2-during'), 40);
    await new Promise((resolve) => setTimeout(resolve, 100));
    await cluster.stop(providerB);
    const duringScaleIn = await traffic;
    const rejected = duringScaleIn.flatMap((outcome) =>
      outcome.status === 'rejected' ? [String(outcome.reason)] : []);
    ensure(
      duringScaleIn.every(isExpectedScaleInOutcome),
      `RM-B2 in-flight request did not finish with a reply or submit timeout: ${rejected.join(' | ')}`
    );
    await cluster.waitForSingleProvider('api-a', providerA.channelEndpoint);
    const beforeAfterA = await getJson<string[]>(providerA.httpUrl, '/evidence');
    const markerAfter = uniqueMarker('rm-b2-after');
    const repliesAfter: ProfileRes[] = [];
    for (let i = 0; i < 20; i += 1) {
      repliesAfter.push(await postJson<ProfileRes>(consumer.httpUrl, '/profile/request', { value: `${markerAfter}-${i}` }));
    }
    ensure(repliesAfter.every((reply) => reply.providerRid === 'api-a'), 'RM-B2 after scale-in should reach api-a only.');
    const afterA = await postJson<string[]>(providerA.httpUrl, '/evidence/wait', { contains: `${markerAfter}-19` });
    const a = countNewEvidence(afterA, beforeAfterA, 'profile-request|rid=api-a', markerAfter);
    ensure(a === 20, 'RM-B2 expected only api-a after scale-in.');
    console.log('scenario RM-B2 passed');
  } finally {
    await cluster.close();
  }
}

function isExpectedScaleInOutcome(outcome: PromiseSettledResult<ProfileRes>): boolean {
  if (outcome.status === 'fulfilled') {
    return outcome.value.providerRid === 'api-a' || outcome.value.providerRid === 'api-b';
  }
  return String(outcome.reason).includes('ZLink async submit timed out.');
}

async function sendContinuously(
  consumerUrl: string,
  marker: string,
  count: number
): Promise<PromiseSettledResult<ProfileRes>[]> {
  const requests: Promise<ProfileRes>[] = [];
  for (let i = 0; i < count; i += 1) {
    requests.push(postJson<ProfileRes>(consumerUrl, '/profile/request', { value: `${marker}-${i}` }));
    await new Promise((resolve) => setTimeout(resolve, 25));
  }
  return await Promise.allSettled(requests);
}
